### ISSUE TITLE
Raw cookies per second is set to a very high value on initial load

### ISSUE THREAD

**Issue by erbkaiser** [erbkaiser/FrozenCookies#19]
Something in the code is causing `Game.cookiesPsRawHighest` (highest raw cookies per second this ascension) to be set to 70.4 quadrillion even in an empty save with everything disabled.
This messes with the stock market as it's based on the highest value ever for that save.

Setting `Game.cookiesPsRawHighest` to zero in the FC loader does not solve this, it seems to happen shortly after load. Nothing is shown in the console.

**Comment by erbkaiser** [comment]
This bug exists (with a different value for cookiesPsRawHighest) in upstream, too, so it's not new to this fork.

That the value _changed_ with my fork might suggest it's something with the efficiency calculations as those run regardless of any other functions being active, and with my fork more upgrades are being tracked.

**Comment by Darkroman** [comment]
Even the OG FC by Icehawk78 gives a number of 8.3 quadrillion cookiesPsRawHighest. I believe I have pinpointed through  tedious debugging and stepping that it happens upon the line var reverseFunctions = upgradeToggle(current); in the function upgradeStats(recalculate). When I step over next function call on that line, the cookiesPs value itself changes constantly from 0.5 to 1 to 2 to 2.5 to 5 and 10 and I think it just keeps building over time. I'm gonna do individual steps now upon that line and I'll see if I notice where the change exactly occurs in the program.

**Comment by Darkroman** [comment]
Seems to happen within whatever it does in Game.CalculateGains() when it is called in fc_main.js on line 3074 (in your fork). Trying to pinpoint within that call exactly hasn't been easy. More updates coming soon.

**Comment by Darkroman** [comment]
I'm stumped. Been at this for a while now. I don't know if it has to do with how FC is just built (I'm not very good in JS I'll be honest). I've gotten it down to where I add breakpoints on lines 3074 in your fork, and 5078 within the actual game code. I was watching for these variables:
Game.cookiesPsRawHighest
Game.cookiesPs
baseCPS()
cpsNew
me.storedTotalCps
me.storedCps
me.amount
me.cps(me)

Perhaps you'll have better luck. It might have something to do with the upgradeStats(recalculate) function as well. My brain is shot for the moment.

**Comment by erbkaiser** [comment]
Good research.
I suspect that because function upgradeToggle (line 3009) actually does a reverse lookup of what it needs to unlock the buildings and upgrades it causes cookiesPsRawHighest to get set _somewhere_.

It might have to do with vanilla line 15106 onwards, //experimental debugging function that cycles through every owned upgrade, turns it off and on, and lists how much each upgrade is participating to CpS

**Comment by Darkroman** [comment]
So it has to do with FC setting building.amount (for some reason, no idea) and it using that value I believe when iterating through. Reverse function just kinda reverses it back to what it should be. Even in saved games with quintillion prestige and rocking 64 septillion cps, the cookiesPsRawHighest is set slightly higher than it should be. Honestly with how the code was stepping, a good fix would be to somehow do Game.CalculateGains() ONLY AFTER the reverse function has done its job.

Perhaps switch the placement of Game.CalculateGains() to be after return reverseFunctions on line 3074? I would do my own testing but for some reason my fork from yours isn't loading up.

**Comment by Darkroman** [comment]
Scratch that, scratch all of that. We need Icehawk to fix this lol. The way it works is deeply built into FC. It'd practically require a rewrite of how it gets all the information needed for the efficiency and next purchase listings.

**Comment by erbkaiser** [comment]
To test your fork you need to setup pages for it and have it served from darkroman.github.io or your own domain if you have one (or from a local jekyll install). Your frozen_cookies.js is now pointing to https://github.darkroman.com/, which does not exist.

I did a bit of testing with my local web server build and I think you're right, this is so deep in the code t would require a full rewrite of the efficiency calculations, which is beyond my current Javascript skill.

The only workaround I found is to force run a delayed **Game.cookiesPsRawHighest = 0**, about 30s after initial load, but this of course deletes the actual cookiesPsRawHighest for running games.

I am probably going to close this as wontfix as a known issue, since it's just not feasible to do a full rewrite ☹️

**Comment by erbkaiser** [comment]
Labelling this as wontfix and closing. It's tied deeply in how FC works and would basically require rewriting a huge part of the mod, only to lose the accurate efficiency calculations.

**Comment by erbkaiser** [comment]
Root issue: https://github.com/Icehawk78/FrozenCookies/issues/217

**Comment by erbkaiser** [comment]
For anyone still tracking this: I believe I have finally fixed it.

**Comment by erbkaiser** [comment]
Tested in browser and Steam.
Old saves should be fixed after ascending.
